ProjectSource/app-library/src/test/java/com/kayako/sdk/android/k5/messenger/attachmentpreview/AttachmentPreviewFragmentTest.java

```diff
@@ -0,0 +1,373 @@
+
+})
+public class AttachmentPreviewFragmentTest {
+    @Rule
+    public ExpectedException expectedException = ExpectedException.none();
+    private AttachmentPreviewFragment underTest;
+    @Mock
+    private View loadingView;
+    @Mock
+    private View mRoot;
+    @Mock
+    private FileAttachmentDownloadHelper mFileAttachmentDownloadHelper;
+    @Mock
+    private DownloadAttachment mDownloadAttachment;
+    @Mock
+    private LayoutInflater inflater;
+    @Mock
+    private View view;
+    @Mock
+    private Bundle bundle;
+    @Mock
+    private FragmentHostCallback mHost;
+    @Mock
+    private Activity mActivity;
+    @Mock
+    private FragmentActivity activity;
+    @Mock
+    private View sendButton;
+    @Mock
+    private View optionsButton;
+    @Mock
+    private View attachmentDetails;
+    @Mock
+    private TextView textView;
+    @Mock
+    private TextView attachmentTime;
+    @Mock
+    private Auth auth;
+
+    @Before
+    public void setUp() {
+        MockitoAnnotations.initMocks(this);
+        staticMock();
+        underTest = spy(new AttachmentPreviewFragment());
+
+        setInternalState(underTest, "mRoot", mRoot);
+        setInternalState(underTest, "mHost", mHost);
+        setInternalState(underTest, "mDownloadAttachment", mDownloadAttachment);
+        setInternalState(underTest, "mFileAttachmentDownloadHelper", mFileAttachmentDownloadHelper);
+        setInternalState(mHost, "mActivity", mActivity);
+
+        doReturn(activity).when(underTest).getActivity();
+    }
+
+    @Test
+    public void givenOnCreateViewThenInflaterShouldInflate() {
+        underTest.onCreateView(inflater, null, null);
+        verify(inflater).inflate(R.layout.ko__fragment_attachment_preview, null, false);
+    }
+
+    @Test
+    public void givenOnViewCreatedThenShouldSetOnClickListener() {
+        View mockedView = mock(View.class);
+        doReturn(mockedView).when(mRoot).findViewById(R.id.ko__button_exit);
+        underTest.onViewCreated(view, bundle);
+        verify(mockedView).setOnClickListener(any(View.OnClickListener.class));
+    }
+
+    @Test
+    public void givenOnActivityCreatedWhenActivityNotKayakoAttachmentPreviewActivityThenThrowIllegalStateException() {
+        expectedException.expect(Exception.class);
+        expectedException.expectMessage("This fragment is meant to be used with KayakoAttachmentPreviewActivity");
+        FragmentActivity activity = mock(FragmentActivity.class);
+        when(underTest.getActivity()).thenReturn(eq(activity));
+        underTest.onActivityCreated(bundle);
+    }
+
+    @Test
+    public void givenOnActivityCreatedWhenImageUrlNotNullThenSetOnClickListener() {
+        setUpForOnActivityCreated();
+        String imageUrl = "imageUrl";
+        String filePath = null;
+        boolean showSendButton = true;
+        View exitButton = mock(View.class);
+        when(bundle.getString(anyString())).thenReturn(imageUrl, filePath, "sessionId", "userAgent");
```

Comment:
⚠️  This snippet of code uses so-called magic strings and sometimes is difficult to infer their goal in the application. It is a good practice to "extract" magic strings and provide meaningful names for them. In this case, consider extracting them into class-level symbolic constants. This approach increases maintainability (single point of modification) and debugging, avoid duplication and decrease verbosity.  

**[More details](https://confluence.devfactory.com/x/SI8-Ew)**
          
## The solution in this case

```Java  
    private static final String USER_AGENT = "userAgent";
    private static final String SESSION_ID = "sessionId";
    // replace 3 usages of these literals with the new constant
```
 

<a id="1b65a140-4faa-11e8-b7bf-a93e6485754a" href="none" codereview-uuid-reserved-tag />